endpoint.go

```diff
@@ -515,11 +514,13 @@ func (ep *endpoint) sbJoin(sb *sandbox, options ...EndpointOption) (err error) {
 	extEp := sb.getGatewayEndpoint()
 
 	sb.Lock()
-	heap.Push(&sb.endpoints, ep)
+	sb.addEndpoint(ep)
 	sb.Unlock()
 	defer func() {
 		if err != nil {
+			sb.Lock()
 			sb.removeEndpoint(ep)
+			sb.Unlock()
 		}
```

Comment:
Rather than removing the sandbox locking from sandbox.removeEndpoint() why not have sandbox.addEndpoint() lock like the rest of the sandbox methods that iterate over the endpoint set?  Is there a reason for add/remove to be separated out?   If not, please put the locking within the method.